Triangulation_2/include/CGAL/Triangulation_2/internal/Polyline_constraint_hierarchy_2.h

```diff
@@ -1055,189 +1192,162 @@ split_constraint(T va, T vb, T vc){
-  hcl = (*sc_iter).second;
-  return true;
-}
-
-template <class T, class Compare, class Point>
-inline
-bool
-Polyline_constraint_hierarchy_2<T,Compare,Point>::
-get_contexts(T va, T vb,
-             Context_iterator& ctxt,
-             Context_iterator& past) const
-{
-  Context_list* hcl;
-  if (!get_contexts(va,vb,hcl)) return false;
-  ctxt = hcl->begin();
-  past = hcl->end();
-  return true;
+  Sc_iterator sc_iter = find_contexts(va, vb);
+  if(sc_iter == contexts_not_found())
+    return nullptr;
+  else
+    return sc_iter->second;
 }
 
-
-
 template <class T, class Compare, class Point>
 inline
-typename Polyline_constraint_hierarchy_2<T,Compare,Point>::Vertex_it
-Polyline_constraint_hierarchy_2<T,Compare,Point>::
-get_pos(T va, T vb) const
-  //return pos in the first context
+auto Polyline_constraint_hierarchy_2<T,Compare,Point>::
+contexts(T va, T vb) const -> Iterator_range<Context_iterator>
 {
-    return (*sc_to_c_map.find(make_edge(va,vb))).second->begin().pos;
-}
-
-template <class T, class Compare, class Point>
-void
-Polyline_constraint_hierarchy_2<T,Compare,Point>::
-oriented_end(T va, T vb, T& vc) const
-{
-  Context_iterator ctxt, past;
-  if(!get_contexts(va,vb, ctxt, past) ) CGAL_assertion(false);
-  if(*(ctxt->pos) == va)
-    vc = ctxt->enclosing->back();
-  else
-    vc = ctxt->enclosing->front();
+  Context_list* hcl = get_context_list(va,vb);
+  if (nullptr == hcl) return {};
+  else return {hcl->begin(), hcl->end()};
 }
 
 
 template <class T, class Compare, class Point>
 void
 Polyline_constraint_hierarchy_2<T,Compare,Point>::
-print() const
+print(std::ostream& os) const
 {
-  C_iterator hcit;
-  std::map<T,int>  vertex_num;
+  std::map<Vertex_handle,int>  vertex_num_map;
   int num = 0;
-  for(hcit = c_begin(); hcit != c_end();  hcit++) {
-    Constraint_id cid = (*hcit);
-    Vertex_it vit =cid.vl_ptr()->skip_begin(), end = cid.vl_ptr()->skip_end();
-    for (;vit != end; vit++){
-      num ++;
-      vertex_num.insert(std::make_pair((*vit), num));
+  for(const auto& cid : constraints()) {
+    for (const auto& node : cid.elements()){
+      vertex_num_map.emplace(node.vertex(), ++num);
     }
   }
-//  typename std::map<T,int>::iterator vnit = vertex_num.begin();
-//  for(; vnit != vertex_num.end(); vnit++) {
-//    vnit->second = ++num;
-//    std::cerr << "vertex num " << num  << " " << vnit->first->point()
-//              << std::endl;
-//  }
-
-  C_iterator cit=c_begin();
-  Sc_iterator scit=sc_begin();
-
-  for(; cit != c_end();  cit++){
-    std::cout << std::endl ;
-    std::cout << "constraint " ;
-    std::cout << cit->vl_ptr();
-    std::cout << "  subconstraints " ;
-    Vertex_it vit = (*cit).vl_ptr()->skip_begin(), end = (*cit).vl_ptr()->skip_end();
-    for(; vit != end; vit++){
-      std::cout << vertex_num[*vit]  <<" ";
+
+  auto disp_vertex = [&vertex_num_map](Vertex_handle v) {
+    return CGAL::IO::oformat(
+        [v, &vertex_num_map](auto& os) -> decltype(os)& {
```

Comment:
```suggestion
        [v](auto& os) -> decltype(os)& {
```

Reply:
I must reject this suggestion, because `vertex_num_map` is used, line 1310, in case the vertex type does not have timestamps. I will add a `[[maybe_unused]]`.